Add GitHub Actions workflow

owenvoke opened the pull request:
I think, as mentioned [in Gitter](https://gitter.im/tldr-pages/tldr?at=5ebc6dd920eaac185309c6cc), it would be good to move to GitHub Actions for our workflows instead of Travis.

I have created a workflow file that uses the [Setup Elixir](https://github.com/actions/setup-elixir) action however it seems to be failing on the `mix deps.get` call:
https://github.com/owenvoke/extldr/runs/674015656?check_suite_focus=true#step:4:1

```
init terminating in do_boot ()
{"init terminating in do_boot",{undef,[{elixir,start_cli,[],[]},{init,start_em,1,[]},{init,do_boot,3,[]}]}}

Crash dump is being written to: erl_crash.dump...done
##[error]Process completed with exit code 1.
```

I haven't really got any experience with Elixir, so @ivanhercaz, do you have any ideas? 🤔 It seemed to be complaining previously about unsupported versions, so perhaps this is due to a version missmatch between OTP and Elixir? 🤷

owenvoke committed:
Add GitHub Actions workflow

ivanhercaz commented:
Thank you very much for the initiative @owenvoke. I will take a look at this as soon as possible to perform the migration to GitHub Actions.

ivanhercaz commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,41 @@
+      fail-fast: true
+      matrix:
+        elixir: [1.9.x, 1.8.x, 1.7.x]
+        otp: [22.3, 21.2, 19.x]
```
```suggestion
        otp: [22.3, 21.2, 19.2]
```
After check the log of the build, I think the reason of the fail is the OTP v19.x in the `otp` matrix. In the README of [actions/setup-elixir](https://github.com/actions/setup-elixir) there is a note that seems relevant for this case:

> Note The OTP release version specification is relatively complex. For best results, we recommend specifying exact OTP and Elixir versions. However, values like 22.x are also accepted, and we attempt to resolve them according to semantic versioning rules.

So let's try with the version 19.2, although this one doesn't seem available on Ubuntu Packages repository. I introduce the test for Erlang 19 it is the newer version supported on Debian 9 (Stretch), and I was interested in this testing personally although I use a newer version of Debian and have Erlang installed by another way, not the Debian Package.

In the case the build will not run after this change, it will probably be necessary to remove this test due it would be because there is not Erlang 19 on Ubuntu Packages repository.

ivanhercaz replied:
@owenvoke, I am going to perform the suggestion to check if it works, and if not works, begin to think about an alternative or if it is necessary to remove this version from the tests.

**PS**. You were fastest than me! :tada:

owenvoke replied:
I did originally try with v17.3 (which is one of the versions used in Travis), but that failed with an incompatible warning. 🤷 I think the action was updated recently to use hex.pm for installs, so that should allow the older versions to be installed. However... Looks like it still failed. 😬
https://github.com/owenvoke/extldr/runs/674285141?check_suite_focus=true

ivanhercaz replied:
I am going to review it deeply. I will notice you any new about it.

owenvoke committed:
Update OTP to specific version

Co-authored-by: Iván Hernández Cazorla <ivan@ivanhercaz.com>

ivanhercaz committed:
github-ci: update OTP versions

With this change the OTP version 19.2 is changed to 18.3 to check if this one, available on the Ubuntu Packages repository, works as expected.

ivanhercaz commented:
Excuse me @owenvoke, I edited the configuration to check quickly if it works. [18.3 is available](https://packages.ubuntu.com/xenial/erlang) for Ubuntu. I would swear that 17.3 was already available, due it works with Travis CI which also used Ubuntu repositories, but it seems it isn't the case. Let's check what happens with 18.3.

owenvoke commented:
No problem 👌 You can edit it as much as you want, and we can squash it when merging. 👍

It seems to be running the actions on my branch (I assume because this is a draft PR, I'll make it ready for review).
https://github.com/owenvoke/extldr/actions

Looks like it's still failing. For Elixir 1.7, with OTP 18.3, it gets:
> unsupported Erlang/OTP version, expected Erlang/OTP 19+

For Elixir 1.8, with OTP 18.3, it gets:
> {"init terminating in do_boot",{undef,[{elixir,start_cli,[],[]},{init,start_it,1,[]},{init,start_em,1,[]}]}}

ivanhercaz commented:
> No problem  You can edit it as much as you want, and we can squash it when merging.

Blessed squassing! Am I dreaming or does it works fine now?

**PS.** No, [it didn't works](https://github.com/owenvoke/extldr/runs/674317002?check_suite_focus=true).

ivanhercaz commented:
> It seems to be running the actions on my branch (I assume because this is a draft PR, I'll make it ready for review).

This might be because it isn't already integrated in the main repository.

> Looks like it's still failing. For Elixir 1.7, with OTP 18.3, it gets:
> > unsupported Erlang/OTP version, expected Erlang/OTP 19+

> For Elixir 1.8, with OTP 18.3, it gets:
> > {"init terminating in do_boot",{undef,[{elixir,start_cli,[],[]},{init,start_it,1,[]},{init,start_em,1,[]}]}}

I am going to work on this.

ivanhercaz commented:
@owenvoke I think I have the solution! Yes, as you said, the problem should be compatibility between Elixir and Erlang. I am checking [this table of compatibility between them](https://hexdocs.pm/elixir/compatibility-and-deprecations.html#compatibility-between-elixir-and-erlang-otp).

ivanhercaz commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,41 @@
+      fail-fast: true
+      matrix:
+        elixir: [1.9.x, 1.8.x, 1.7.x]
+        otp: [22.3, 21.2, 18.3]
```
```suggestion
        otp: [22.3, 21.2, 20.2]
```

Very important! Erlang/OTP is not supported by Elixir 1.7. The support of Elixir 1.7 is for Erlang versions between the v19 and the v22. The latest Elixir that supports Erlang/OTP v18 was Elixir 1.5!

So I have two ideas:

  - Remove the 18.3
  - Replace with a supported version of Erlang/OTP available on Ubuntu Packages repository.

In this suggestion I suggest to apply the second option, due it is available a v20.2 version of Erlang/OTP.

ivanhercaz committed:
github-ci: update OTP versions

owenvoke commented:
I think Coveralls is just failing because my fork hasn't been set up 👍 Looks like you've fixed the issue though!

ivanhercaz commented:
Current status:
- **CI - Elixir 1.9.x - OTP 22.3**: failed to upload report to Coveralls.
- **CI - Elixir 1.9.x - OTP 21.2**: failed to upload report to Coveralls.
- **CI - Elixir 1.9.x - OTP 20.2**: failed to upload report to Coveralls.
- **CI - Elixir 1.8.x - OTP 22.3**: failed to upload report to Coveralls.
- **...**: the rest have the same error, less one that was cancelled.

ivanhercaz commented:
> I think Coveralls is just failing because my fork hasn't been set up  Looks like you've fixed the issue though!

Yes! It seems the issue is fixed and the current problem with Coveralls might be caused due Coveralls is configured with this repository and no the forks.

What do you think?! Should we proceed to merge it?

ivanhercaz commented:
Just a detail, I think we should remove Travis when we will be sure that GitHub Actions works fine, so Travis would be removed in a another PR. Just to be sure everything works as expected and that we don't remove the Travis configuration before.

**PS**. I will merge it after lunch to be sure that then everything works fine!

owenvoke commented:
I think that sounds good to me! 👌

ivanhercaz commented:
@owenvoke, merged! Now [it is running](https://github.com/tldr-pages/extldr/actions/runs/104695716) by the first in this repository.
